src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,282 @@
+
+#ifndef RIPPLE_CORE_SQLINTERFACE_H_INCLUDED
+#define RIPPLE_CORE_SQLINTERFACE_H_INCLUDED
+
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/app/misc/NetworkOPs.h>
+#include <ripple/app/misc/Transaction.h>
+#include <ripple/core/Config.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/peerfinder/impl/Store.h>
+#include <boost/filesystem.hpp>
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+
+namespace ripple {
+
+class SQLInterface
+{
+public:
+    enum class DatabaseType { Ledger, Transaction, Last };
+
+    enum class TableType { Ledgers, Transactions, AccountTransactions, Last };
+
+    typedef std::pair<LedgerIndex, uint8_t> DatabaseIndex;
+
+    static std::string
+    tableName(TableType type)
+    {
+        /* need to modify switch statement if added new value to TableType */
+        switch (type)
+        {
+            case TableType::Ledgers:
+                return "Ledgers";
+            case TableType::Transactions:
+                return "Transactions";
+            case TableType::AccountTransactions:
+                return "AccountTransactions";
+            default:
+                break;
+        }
+        return "Unknown";
+    }
+
+    struct RowsMinMax {
+        int numberOfRows;
+        int minLedgerSequence;
+        int maxLedgerSequence;
+    };
+
+    struct Hashes {
+        uint256 ledgerHash;
+        uint256 parentHash;
+    };
+
+    static std::unique_ptr<SQLInterface>
+    init(Application& app, Config const& config);
+
+    virtual bool
+    makeLedgerDBs(
+        Config const& config,
+        DatabaseCon::Setup const& setup,
+        DatabaseCon::CheckpointerSetup const& checkpointerSetup) = 0;
+
+    virtual void
+    addDBs(LedgerIndex shardIndex, DatabaseCon* tx, DatabaseCon* lgr) = 0;
+
+    virtual void
+    removeDBs(LedgerIndex shardIndex) = 0;
+
+    virtual ~SQLInterface() = default;
+
+    virtual boost::optional<LedgerIndex>
+    getMinLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual boost::optional<LedgerIndex>
+    getMaxLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual void
+    deleteByLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual void
+    deleteBeforeLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual int
+    getRows(DatabaseType dbtype, TableType type) = 0;
+
+    virtual struct RowsMinMax
+    getRowsMinMax(DatabaseType dbtype, TableType type) = 0;
+
+    virtual bool
+    saveValidatedLedger(
+        Application& app,
+        std::shared_ptr<Ledger const> const& ledger,
+        bool current) = 0;
+
+    virtual std::optional<LedgerInfo>
+    loadLedgerInfoByIndex(
+        LedgerIndex ledgerSeq,
+        beast::Journal j) = 0;
+
+    virtual std::optional<LedgerInfo>
+    loadOldestLedgerInfo(
+        beast::Journal j) = 0;
```

Comment:
Can we remove the `beast::Journal` parameter from these functions? The journal can just be a member variable of the derived classes.

Reply:
Done. App parameters also removed.